### PR TITLE
Fix infinite loops and performance of NPC catching up on moves from time spent outside reality bubble

### LINKED ISSUE #38630: Game Freeze upon approaching of base.
# Describe the bug

When approaching the base to the east the game freezes. Not a crash, and no inputs are possible.

# Steps To Reproduce

1. Approach the base in the east of the current player location. 
2. When approaching from the west or south the NPC gets put in stasis first and the game enters debug mode. When approaching from the north the game simply freezes. 
3. When the NPC gets put into stasis they can be woken up with no problem, but getting closer to the building freezes the game. Teleportation into the building leads to an immediate freeze. 

# Expected behavior

That the game process continuous on as normal. 

# Versions and configuration

- Game Version: 0.D-12917-gdccc2bf [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Magiclysm [magiclysm],
    Garden Pots [growable-pots],
    Makeshift Items Mod [makeshift],
    Medieval and Historic Classes and Shields [Medieval_Stuff],
    More Survival Tools [More_Survival_Tools],
    Alternative Map Key [alt_map_key],
    Mutant NPCs [mutant_npcs],
    My Sweet Cataclysm [my_sweet_cataclysm],
    More Locations [more_locations],
    Folding Parts pack [deoxymod],
    Sleep Deprivation [sleepdeprivation],
    Stats Through Skills [StatsThroughSkills],
    SpeedyDex [speedydex]
]

# Additional context

The .zip file is too big (45 MB) I put it into dropbox. https://www.dropbox.com/s/0bse6jyrewdt49f/Nauvoo-zip.zip?dl=0

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Bugfixes "Fix infinite loops and performance of NPC catching up on moves from time spent outside reality bubble"```

#### Purpose of change
Fixes #38630 

#### Describe the solution
Two things were happening.

1. The NPCs had been outside reality bubble for around 27 hours, and had over 10 million moves stored up to use all at once, around 4 of them or so. 
They had been assigned a chop tree activity, in a large area, so they had to chop down all those trees, then they all ran the tidy_up activity, which was structured in a bad way, especially considering that the sheer AMOUNT of logs and sticks to tidy up,and the HUGE size of the unsorted zones present in the attached save i nthe linked issue.
The previous code would iterate over every point of an unsorted zone in a large radius, and every tile and every item in every tile in a large radius every time they scanned for things to tidy up.
I calculated it might have taken around 6 hours for the game to finish processing it on my laptop :)

Ive got it down to around 5 seconds now in the extreme case of this save having huge zones and a lot of moves to catch up on, still a noticeable blip, but no longer a terrible one - did this by restructuring the map scanning code in the tidy up activity.

We may want to consider some popup text wiht a spinning bar "processing NPC activities" or something ( like we do for player activity processing )

and 2. The second bug here was an NPC doing vehicle deconstructing , triggered the infinite loop counter and reset and popped up a debugmsg.

This was caused due to <deep breath>

when the multi-activity types set a destination, then return, they dont subtract moves, this isnt a problem, because normally, theyd go and do the activity, subtract some moves, use them all up, and exit out of the main game loop, and the game NPC infinite loop counter never goes above 1.   
   But if they have 10 million moves stockpiled up, they will use a million of them going about their business doing lots of different activities, but each time they set a destination and return without using moves, the counter goes up by 1, and because its still within that same game loop the infinite loop counter eventually trips.
I've fixed this by detecting this in the npc::do_player_activity() function and subtracting a nominal move amount to satisfy the counter.

Also some associated cleanup of redundant code and extra safeguarding in the activity code;
By converting it to a function that has a scan_only parameter, and returns a bool, an NPC witha  huge backlog of moves, can potentially just scan once and cancel the activity and zero their moves if need be, instead of ( in some cases ) constantly scanning and trying to process , until their moves eventually hit zero naturally.

#### Describe alternatives you've considered
NPCs drive me to drink, I swear.

#### Testing
Attached save in issue - loaded it up, walked East.
Before : game froze ( and woulda froze for a very very very long time )
after : game hitched for a few seconds but was otherwise ok, and all activity moves backlogs were completed sucesfully.

Walked South-East to vehicle deconstrructing NPC.
Before : NPC went into a coma
After : NPC finished their activity after a brief bit of lag.

#### Additional context
N/A

### CODE REVIEW

**Comment by davidpwbrown** [comment]
Actually just tested and found another problem... not quite ready.

**Comment by davidpwbrown** [comment]
Actually, the 5 second pause was my fault, I was scanning for paths unnecessarily, got the blip down to nearly instant, much better than 6 hours!

File: src/activity_item_handling.cpp [hunk 1]
```diff
@@ -2493,10 +2512,10 @@ static std::unordered_set<tripoint> generic_multi_activity_locations( player &p,
 }
 
 /** Check if this activity can not be done immediately because it has some requirements */
-/** Returns true if this multi activity may be processed further */
-static bool generic_multi_activity_check_requirement( player &p, const activity_id &act_id,
+static int generic_multi_activity_check_requirement( player &p, const activity_id &act_id,
```
Comment (AMurkin) [inline src/activity_item_handling.cpp]: Return type (int) should be an enum?

File: src/activity_handlers.h [hunk 2]
```diff
@@ -90,7 +90,7 @@ int butcher_time_to_cut( const player &u, const item &corpse_item, butcher_type
 void activity_on_turn_drop();
 void activity_on_turn_move_items( player_activity &act, player &p );
 void activity_on_turn_move_loot( player_activity &act, player &p );
-void generic_multi_activity_handler( player_activity &act, player &p );
+bool generic_multi_activity_handler( player_activity &act, player &p, bool check_only = false );
```
Comment (AMurkin) [inline src/activity_handlers.h]: What does the return value mean?

File: src/activity_item_handling.cpp (hunk 1 above)
Comment (davidpwbrown) [reply src/activity_item_handling.cpp]: changed to enum return type

File: src/activity_handlers.h (hunk 2 above)
Comment (davidpwbrown) [reply src/activity_handlers.h]: added comment

File: src/activity_item_handling.cpp [hunk 3]
```diff
@@ -2493,10 +2512,10 @@ static std::unordered_set<tripoint> generic_multi_activity_locations( player &p,
 }
 
 /** Check if this activity can not be done immediately because it has some requirements */
-/** Returns true if this multi activity may be processed further */
-static bool generic_multi_activity_check_requirement( player &p, const activity_id &act_id,
+static requirement_check_result generic_multi_activity_check_requirement( player &p, const activity_id &act_id,
         activity_reason_info &act_info,
```
Comment (kevingranade) [inline src/activity_item_handling.cpp]:
```suggestion
static requirement_check_result generic_multi_activity_check_requirement( player &p,
        const activity_id &act_id, activity_reason_info &act_info,
```

File: src/activity_item_handling.cpp [hunk 4]
```diff
@@ -2787,41 +2810,41 @@ void generic_multi_activity_handler( player_activity &act, player &p )
...
-        if( !generic_multi_activity_check_requirement( p, activity_to_restore, act_info, src, src_loc,
-                src_set ) ) {
-            return;
-        }
-        //skip this location if doing activity here is impossible
-        if( !act_info.can_do ) {
+        // see activity_handlers.h enum for requirement_check_result
+        const requirement_check_result req_res = generic_multi_activity_check_requirement( p, activity_to_restore, act_info, src,
+                            src_loc,
+                            src_set, check_only );
```
Comment (kevingranade) [inline src/activity_item_handling.cpp]:
```suggestion
        const requirement_check_result req_res = generic_multi_activity_check_requirement( p,
                            activity_to_restore, act_info, src, src_loc, src_set, check_only );
```

**Pull request merged**